walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,135 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        // Create full (mandatory + optional) attendee list
+        Collection<String> fullAttendees = new ArrayList<>();
+        fullAttendees.addAll(request.getAttendees());
+        fullAttendees.addAll(request.getOptionalAttendees());
+
+        MeetingRequest fullRequest = new MeetingRequest(fullAttendees, request.getDuration());
+        Collection<TimeRange> resultWithOptionals = queryHelper(events, fullRequest);
+
+        // If any TimeRanges are available with optionals include, return those,
+        // Otherwise exculde optionals
+        // Notable case: When there are no mandatory attendees, optionals are treated as mandatory and returned.
+        if (request.getAttendees().isEmpty() || !resultWithOptionals.isEmpty()) {
+            return resultWithOptionals;
+        } else {
+            return queryHelper(events, request);
+        }
+    }
+
+
+    public Collection<TimeRange> queryHelper(Collection<Event> events, MeetingRequest request) {
+
+        // Remove events that do not share in required attendees.
+        ArrayList<Event> eventList = new ArrayList<>(events);
+        AttendeeHelper attendees = new AttendeeHelper(request.getAttendees());
+        
+        Iterator<Event> eventIt = eventList.iterator();
+        while (eventIt.hasNext()) {
+            if (!attendees.eventHasAttendees(eventIt.next())) {
+                eventIt.remove();
+            }
+        }
+        ArrayList<TimeRange> eventTimeRanges = new ArrayList<>();
+        for (Event event : eventList) {
+            eventTimeRanges.add(event.getWhen());
+        }
```

Comment:
It seems like you could avoid unnecessary copying here by combining the 2 loops. Instead of first iterating over the events and removing events, and then going over the result and adding the remaining events to a different ArrayList, maybe you could do something like:

1. Create ArrayList `eventTimeRanges`
2. For each `event` in `events`
    - if `eventHasAttendees` then add it to `eventTimeRanges`